### ISSUE TITLE
Can I use variable for Field in set_value?

### ISSUE THREAD

**Issue by BRC555** [hongyuanjia/eplusr#21]
For my case, I can change the value of a field like this:

model6012Try$"Coil:Cooling:DX:SingleSpeed"$Thermal_Zone_ComputerRoom1_Coil_Cooling_DX_Single_Speed$set_value("Gross_Rated_Cooling_COP"=1.2)
Or
model6012Try$"Coil:Cooling:DX:SingleSpeed"$Thermal_Zone_ComputerRoom1_Coil_Cooling_DX_Single_Speed$set_value(Gross_Rated_Cooling_COP=2.2)

But I cannot do like this in a loop which automatically get the field name from a csv file:
AAA1="Gross_Rated_Cooling_COP";
model6012Try$"Coil:Cooling:DX:SingleSpeed"$Thermal_Zone_ComputerRoom1_Coil_Cooling_DX_Single_Speed$set_value(AAA1=3.2);

Could I know any solution for my case?
Thanks a lot.

**Comment by hongyuanjia** [comment]
There is no non-standard evaluation (NSE) feature implemented in eplusr yet.
`AAA1` is treated as normal IDF field name. But by using rlang package, I think you could achieve what you want. Considering this below:

```r
library(rlang)
library(eplusr)

AAA1 <- "Gross_Rated_Cooling_COP";
model6012Try$"Coil:Cooling:DX:SingleSpeed"$Thermal_Zone_ComputerRoom1_Coil_Cooling_DX_Single_Speed$set_value(!!AAA1:=3.2);
```

Please note the using of `!!` and `:=` which come from rlang package. I haven't try on my side yet as I can't access to my computer right now.

I am currently refactoring model editing functionalities to make `$set_value()` more programming friendly. Stay tuned for next release.

**Comment by BRC555** [comment]
Dear Hongyuan:

Thanks a lot for your kind help!
I tried your way, but there is some error:
[cid:image001.png@01D45E66.E25D56B0]

I tried this way:
aaa="Coil:Cooling:DX:SingleSpeed";
aaa2=paste('"', aaa, '"', sep = "", collapse = ", ");
ObNaChange2= "Thermal_Zone_ComputerRoom1_Coil_Cooling_DX_Single_Speed";
FieldRealName="Gross_Rated_Cooling_COP";
OpResult=99;

eval(parse(text=paste0('model6012Try$',aaa2,'$',ObNaChange2,'$set_value(',FieldRealName,'=',OpResult,')')))
But sometime is OK and sometimes there is error like this:

[cid:image003.png@01D45E67.84295250]

Do you have other suggestions?

Our group is trying to use R to optimize a bunch of parameters values in EnergyPlus. So I am setting the variables to get and send the values directly in the optimization loop.


Thanks,

Borui CUI, Ph.D.
R&D Associate Staff, Oak Ridge National Laboratory
Oak Ridge, TN 37831-6324
Tel: (865) 241-3773 (Office), (865) 299-8596 (Cell)
cuib@ornl.gov<mailto:cuib@ornl.gov>

**Comment by BRC555** [comment]
Thanks for your kind help! 
These two are the missed pictures in the last section:
![capture3](https://user-images.githubusercontent.com/43863301/46672692-1b6df200-cba6-11e8-8d00-778853803e78.JPG)
![capture4](https://user-images.githubusercontent.com/43863301/46672697-1e68e280-cba6-11e8-9b8f-3934a5559f56.JPG)

**Comment by BRC555** [comment]
I think there are some problems to use $set_value if some variables are used to replace the class, object or field.
Can u suggest any other alternative ways to set new value for a specific field in a object in a class， e.g. changing value in the "Default" column of the following screen shot？
![capture5](https://user-images.githubusercontent.com/43863301/46673175-2ecd8d00-cba7-11e8-8267-e7dcd81ef915.JPG)

Thanks a lot.

**Comment by hongyuanjia** [comment]
Thanks for reporting that.

> Thanks for your kind help!
> These two are the missed pictures in the last section:
> ![capture3](https://user-images.githubusercontent.com/43863301/46672692-1b6df200-cba6-11e8-8d00-778853803e78.JPG)
> ![capture4](https://user-images.githubusercontent.com/43863301/46672697-1e68e280-cba6-11e8-9b8f-3934a5559f56.JPG)

There are probably some bugs in `$set_value()`. It would be great if you could send me the debugging info by running `traceback()` after each error occurred. That will make it much easier for me to find where the problem was.


---

> I think there are some problems to use $set_value if some variables are used to replace the class, object or field.
> Can u suggest any other alternative ways to set new value for a specific field in a object in a class， e.g. changing value in the "Default" column of the following screen shot？
> ![capture5](https://user-images.githubusercontent.com/43863301/46673175-2ecd8d00-cba7-11e8-8267-e7dcd81ef915.JPG)
> 
> Thanks a lot.

Could you please send me the IDF file if there is no confidential concern? Then I can try on my side.

As for alternative ways to set values, you can use `$set_object()` in `Idf` class. As I don't have the model, just using an example file here to demonstrate:

```r

library(eplusr)
library(data.table) # for reading input data

# taking this as an example
path_idf <- file.path(eplus_config(8.8)$dir, "ExampleFiles", "HospitalLowEnergy.idf")

idf <- read_idf(path_idf)

# read input data
input <- data.table::fread(
  "
  index , class             , object                                     , field , default , new_value
  1     , Coil:Heating:Fuel , Floor 1 Cafe Heat Pump Sup Heat Coil       , 4     , 0.8     , 0.7
  2     , Coil:Heating:Fuel , Floor 1 Clean Heat Pump Sup Heat Coil      , 4     , 0.8     , 0.6
  3     , Coil:Heating:Fuel , Floor 1 Conference Heat Pump Sup Heat Coil , 4     , 0.8     , 0.5
  4     , Coil:Heating:Fuel , Floor 1 Dining Heat Pump Sup Heat Coil     , 4     , 0.8     , 0.4
  "
)

# get target field names
input[, `:=`(field_name = idf$definition(class)[[1L]]$field_name(4)), by = index]

# get correct input value format for $set_object()
format_input_value <- function (field_name, value) {
  val <- list(value)
  names(val) <- field_name
  val
}

input[, `:=`(value = list(format_input_value(field_name, new_value))), by = index]

idf$set_object(object = input$object, value = input$value)

# check results
vapply(idf$object(input$object), function (obj) obj$get_value(4), list(1))
# $Floor_1_Cafe_Heat_Pump_Sup_Heat_Coil
# [1] 0.7
# 
# $Floor_1_Clean_Heat_Pump_Sup_Heat_Coil
# [1] 0.6
# 
# $Floor_1_Conference_Heat_Pump_Sup_Heat_Coil
# [1] 0.5
# 
# $Floor_1_Dining_Heat_Pump_Sup_Heat_Coil
# [1] 0.4
# 
```

**Comment by hongyuanjia** [comment]
Also, please make sure to use the develop version as there was a bug in the CRAN version (v0.9.2) when updating value references (see #20).

You can install the develop version by doing:

```r
devtools::install_github("hongyuanjia/eplusr")
```

**Comment by BRC555** [comment]
> Also, please make sure to use the develop version as there was a bug in the CRAN version (v0.9.2) when updating value references (see #20).
> 
> You can install the develop version by doing:
> 
> ```r
> devtools::install_github("hongyuanjia/eplusr")
> ```

Thanks,
I try to run "devtools::install_github("hongyuanjia/eplusr")"; But there is something wrong:
![capture6](https://user-images.githubusercontent.com/43863301/46683176-8a0a7a00-cbbd-11e8-8327-65db8966af69.JPG)

**Comment by hongyuanjia** [comment]
The new version of 0.9.3 has been landed on CRAN. You can install the latest version from CRAN which has already fixed the issue #20.

**Comment by hongyuanjia** [comment]
In version v0.10.0, `$set()` in `Idf` and `IdfObject` class directly support inputs as lists. Close this.
